### DISCUSSION TITLE
Tracking motion of mass particles suspended in a fluid

### DISCUSSION THREAD

**Discussion by sb50nemo** [idaholab/moose#22504] [Q&A Tools] [answered]
Dear MOOSE team, 

I have an array of mass particles suspended in a fluid. Their position is changing with every time step owing to the force field. The transient positions of those particles are stored in a vector format, something like the following: 

$$
\begin{align}
\mathrm{X}[i, t] \\
\mathrm{Y}[i, t] \\
\mathrm{Z}[i, t] \\
\end{align}
$$

where $i$ is the index of the particle and $t$ stands for the timestep. Such a system with 4 particles can be seen in the screenshot below, where the 4 particles change their position with every timestep. The transient positions of these particles will be stored as explained above, $\mathrm{X}[{i =1..4}, t={1..3} ], \mathrm{Y}[{i =1..4}, t={1..3} ], \mathrm{Z}[{i =1..4}, t={1..3} ]$

<img width="956" alt="Screenshot 2022-10-26 at 2 51 54 PM" src="https://user-images.githubusercontent.com/68417883/198032657-69cd9ea9-43aa-4cde-b58a-3253fa106b5c.png">

I need your kind help in visualising the trajectory of these particles as shown in the following screenshot. Could you please help me by suggesting a tool or method to visualise these transient trajectories in MOOSE?

<img width="321" alt="Screenshot 2022-10-26 at 2 52 11 PM" src="https://user-images.githubusercontent.com/68417883/198032707-e26f5ff1-8994-43da-bd32-b25174cc4b37.png">

**Comment by GiudGiud** [comment, marked as answer]
Hello

And you are using the ray tracing module to do that?
Have you tried these outputs if you are:
https://mooseframework.inl.gov/source/outputs/RayTracingExodus.html ?

Guillaume

**Reply by sb50nemo** [reply]
Hello Guillaume, 

Thank you for your response. Actually, `RayTracingExodus` was my preferred choice to output the trajectories. 
But as I have been advised by you and Derek, I need to write a `RayKernel` for changing the ray trajectory in discussion #21283, I am unable to figure out how to implement that. 
I was wondering if you could help me or guide me on how to change the trajectory using a RayKernel. 

I checked this resource for RayKernels (https://mooseframework.inl.gov/syntax/RayKernels/index.html).

````
const Point some_point(1, 2, 3); // must be within _current_elem!
const Point some_direction(1, 0, 0);
changeRayStartDirection(some_point, some_direction);
````

I am puzzled as to how this can be implemented for my current scheme where the transient positions are stored in form of arrays.
I would appreciate your help on this. Thank you.

**Reply by GiudGiud** [reply]
You would have to treat the particles one by one, the ray kernels act on a single ray not multiple. 

These 3 lines would be fine to initialize the trajectory of one particle.

**Reply by lindsayad** [reply]
@oanaoana this made me think of you

**Reply by GiudGiud** [reply]
any update on this?

**Reply by friedmud** [reply]
Pinging again to see if you have an update on this.
